java/ql/lib/semmle/code/java/security/SensitiveResultReceiverQuery.qll

```diff
@@ -17,45 +17,64 @@ private class ResultReceiverSendCall extends MethodAccess {
-  UntrustedResultReceiverConf() { this = "UntrustedResultReceiverConf" }
+private module UntrustedResultReceiverConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node node) { node instanceof RemoteFlowSource }
 
-  override predicate isSource(DataFlow::Node node) { node instanceof RemoteFlowSource }
-
-  override predicate isSink(DataFlow::Node node) {
+  predicate isSink(DataFlow::Node node) {
     node.asExpr() = any(ResultReceiverSendCall c).getReceiver()
   }
 }
 
+private module UntrustedResultReceiverFlow = TaintTracking::Global<UntrustedResultReceiverConfig>;
+
 private predicate untrustedResultReceiverSend(DataFlow::Node src, ResultReceiverSendCall call) {
-  any(UntrustedResultReceiverConf c).hasFlow(src, DataFlow::exprNode(call.getReceiver()))
+  UntrustedResultReceiverFlow::flow(src, DataFlow::exprNode(call.getReceiver()))
 }
 
-private class SensitiveResultReceiverConf extends TaintTracking::Configuration {
-  SensitiveResultReceiverConf() { this = "SensitiveResultReceiverConf" }
-
-  override predicate isSource(DataFlow::Node node) { node.asExpr() instanceof SensitiveExpr }
+module SensitiveResultReceiverConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node node) { node.asExpr() instanceof SensitiveExpr }
 
-  override predicate isSink(DataFlow::Node node) {
+  predicate isSink(DataFlow::Node node) {
     exists(ResultReceiverSendCall call |
       untrustedResultReceiverSend(_, call) and
       node.asExpr() = call.getSentData()
     )
   }
 
-  override predicate allowImplicitRead(DataFlow::Node node, DataFlow::ContentSet c) {
-    super.allowImplicitRead(node, c)
-    or
-    this.isSink(node)
-  }
+  predicate allowImplicitRead(DataFlow::Node n, DataFlow::ContentSet c) { isSink(n) and exists(c) }
 }
 
-/** Holds if there is a path from sensitive data at `src` to a result receiver at `sink`, and the receiver was obtained from an untrusted source `recSrc`. */
-predicate sensitiveResultReceiver(
+module SensitiveResultReceiverFlow = TaintTracking::Global<SensitiveResultReceiverConfig>;
+
+/**
+ * DEPRECATED: Use `isSensitiveResultReceiver` instead.
+ *
+ * Holds if there is a path from sensitive data at `src` to a result receiver at `sink`, and the receiver was obtained from an untrusted source `recSrc`.
+ */
+deprecated predicate sensitiveResultReceiver(
   DataFlow::PathNode src, DataFlow::PathNode sink, DataFlow::Node recSrc
 ) {
-  exists(ResultReceiverSendCall call, SensitiveResultReceiverConf conf |
-    conf.hasFlowPath(src, sink) and
+  exists(
+    ResultReceiverSendCall call, SensitiveResultReceiverFlow::PathNode srrSrc,
+    SensitiveResultReceiverFlow::PathNode srrSink
+  |
+    src.getNode() = srrSrc.getNode() and sink.getNode() = srrSink.getNode()
+  |
+    SensitiveResultReceiverFlow::flowPath(srrSrc, srrSink) and
+    sink.getNode().asExpr() = call.getSentData() and
+    untrustedResultReceiverSend(recSrc, call)
+  )
```

Comment:
I get that this tries to use the new logic even in the old predicate, but I feel (maybe irrationaly) a bit nervous about `src.getNode() = srrSrc.getNode() and sink.getNode() = srrSink.getNode()`. Why don't we leave it as it was (except it's now deprecated), so that we make sure we aren't impacting existing users of this predicate?

Reply:
Agree. The existence of those `DataFlow::PathNode`s rely on the old configuration anyway, and are best linked up by using the associated `hasFlowPath` predciate.

Reply:
Sounds good. I've re-added the old configuration as `deprecated private`.